### DISCUSSION TITLE
HTTP 메시지는 어떻게 구성되어 있을까요?

### DISCUSSION THREAD

**Discussion by alstn113** [2024-woowacourse-study/level-interview#170] [Network]
.

**Comment by Kimprodp** [comment]
**[HTTP Request]**
- Start-line : 요청의 종류를 나타내는 메서드, URI, HTTP 버전으로 구성되어 있다.
- Header : 요청에 대한 추가 정보를 담고 있는 부분으로 클라이언트 및 요청에 대한 다양한 메타 데이터를 포함하며 각 헤더는 "key : value" 형식으로 저장된다.
- Blank Line : 헤더의 끝을 나타낸다.
- Body : HTTP Request가 전송하는 데이터를 담고 있는 부분으로, 전송하는 데이터가 없다면 body 부분은 공백이다. 보통 POST 요청일 경우, HTML 폼 데이터가 포함되어 있다.

**[HTTP Response]**
- Start-line : 응답 상태를 간략하게 나타내는 부분으로 HTTP 버전, Status Code, 상태 메세지로 구성되어 있다.
- Header : 요청의 Header 처럼 응답에 대한 추가 정보를 담고 있는 부분이다. 요청 헤더와 다른 점은 응답 헤더에서만 사용되는 값들이 존재한다.
- Blank Line : 헤더의 끝을 나타낸다.
- Body : 요청된 리소스의 실제 데이터를 포함하며, HTML 페이지, 이미지, JSON 데이터 등 다양한 형식이 될 수 있다. 데이터를 전송할 필요가 없을 경우 body가 비어있게 된다.

**Comment by seunghye218** [comment]
모든 HTTP 메시지는 시작줄로 시작한다. 요청 메시지의 시작줄은 무엇을 해야하는지 말해준다. 응답 메시지의 시작줄은 무슨 일이 일어났는지 말해준다.

HTTP 헤더 필드는 요청과 응답 메시지에 추가 정보를 더한다. 그들은 기본적으 로 이름/값 쌍의 목록이다.

### 엔터티 본문
HTTP 메시지의 세 번째 부분은 선택적인 엔터티 본문이다. 이미지, 비디오, HTML 문서, 소프트웨어 애플리케이션, 신용카드 트랜잭션, 전자우편 등 여러 종류의 디지털 데이터를 실어 나를 수 있다.

### 요청 메시지의 형식

```
<메서드> <요청 URL> <버전> # 시작줄
<헤더> # 헤더

<엔터티 본문> # 본문
```

### 응답 메시지의 형식

시작줄에서만 문법이 다르다.

```
<버전> <상태코드> <사유구절> # 시작줄
<헤더> # 헤더

<엔터티 본문> # 본문
```

### 메서드

클라이언트 측에서 서버가 리소스에 대해 수행해주길 바라는 동작이다. ‘GET’, ‘HEAD’, ‘POST’와 같이 한 단어로 되어 있다. 

### 요청 URL

요청 대상이 되는 리소스를 지칭하는 완전한 URL 혹은 URL의 경로 구성요소다. 완전한 URL이 아닌 URL의 경로 구성요소라고 해도, 클라이언트가 서버와 직접 대화하고 있고 경로 구성요소가 리소스를 가리키는 절대 경로이기만 하면 대체로 문제가 없다. 서버는 URL에서 생략된 호스트/포트가 자신을 가리키는 것으로 간주할것이다.

### 버전

이 메시지에서 사용 중인 HTTP의 버전이다.

### 상태 코드

요청 중에 무엇이 일어났는지 설명하는 세 자리의 숫자다. 각 코드의 첫 번째 자릿수는 상태의 일반적인 분류(‘성공’, ‘에러’ 등)를 나타낸다.

### 사유 구절(reason-phrase)

숫자로 된 상태 코드의 의미를 사람이 이해할 수 있게 설명해주는 짧은 문구로, 상태 코드 이후부터 줄바꿈 문자열까지가 사유 구절이다.

### 헤더들

이름, 콜론(:), 선택적인 공백, 값, CRLF가 순서대로 나타나는 0개 이상의 헤더들. 이 헤더의 목록은 빈 줄(CRLF)로 끝나 헤더 목록의 끝과 엔터티 본문의 시작을 표시한다.

### 엔터티 본문

엔터티 본문은 임의의 데이터 블록을 포함한다. 모든 메시지가 엔터티 본문을 갖는것은 아니므로, 때때로 메시지는 그냥 CRLF으로 끝나게 된다.

**Comment by J-I-H-O** [comment]
**HTTP Request**

- 시작 줄 (start-line)
    - HTTP 메서드, 요청 타겟 URL, 프로토콜 버전
    - 예) `GET localhost:8080/reservations HTTP/1.1`
- 헤더 (header)
    - 요청에 대한 메타 데이터 포함
    - `key:value` 쌍으로 표현
- 빈 줄 (blank line)
    - 요청에 대한 모든 메타 데이터가 전송되었음을 알리는 용도로 빈 줄 삽입
- 본문 (body)
    - 요청과 관련된 본문
    - GET, HEAD 등 과 같이 리소스를 가져오는 요청은 보통 본문이 필요하지 않음

---

**HTTP Response**

- 상태 줄 (status line)
    - 프로토콜 버전, 상태 코드, 상태 코드에 대한 상태 텍스트
    - 예) `HTTP/1.1 404 Not Found.`
- 헤더 (header)
- 빈 줄 (blank line)
- 본문 (body)
    - payload 없이도 요청에 충분히 응답하는 201 Created, 204 No Content와 같은 응답에는 보통 본문이 존재하지 않음

https://developer.mozilla.org/ko/docs/Web/HTTP/Messages

**Comment by 3Juhwan** [comment]
![image](https://github.com/2024-woowacourse-study/level-interview/assets/64410384/29abc285-fc62-4ef9-a946-1e90c761779a)

출처: https://developer.mozilla.org/ko/docs/Web/HTTP/Messages
